File: frontend/src/stores/evolution.ts
/**
 * Evolution store — shared fetch logic for evolution-related pages
 * Replaces direct `request` calls in EvolutionDashboard.vue (audit B5)
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'
import request from '@/api/request'
import type { ChangeType } from '@/types/evolution'

import { useResponseValidation } from '@/validation/useResponseValidation'
import evolutionSchema from '@contracts/schemas/evolution.schema.json'

// PLAN-014: 契约响应校验 (DEV warn 不阻断)
const { validateResponse: validateEvolution } = useResponseValidation()

export interface TrendItem {
  skill_name: string
  trend: string
  confidence: number
  points: number[]
  related_positions: string[]
}

export interface SnapshotEntry {
  id: string
  position_name: string
  snapshot_date: string
  required_skills: unknown[]
  preferred_skills?: unknown[]
  source_count: number
}

export interface ChangelogEntry {
  id: string
  skill_name: string
  change_type: ChangeType
 // 字段对齐后端 ChangelogEntry schema（Pydantic ↔ openapi.yaml ↔ evolution.schema.json）
  trust_score: number
  confidence: number
  created_at: string
 // Optional fields returned by the backend schema / used in UI
  position_name?: string
  old_proficiency?: string | null
  new_proficiency?: string | null
  old_requirement?: string | null
  new_requirement?: string | null
  date?: string
  description?: string
 // 10-03 contract sync: status/written_back/evidence_json ( 证据链路)
  status?: 'pending' | 'approved' | 'rejected'
  written_back?: boolean
  evidence_json?: Record<string, unknown>
}

// 10-03: KPI row state — zeros by default, filled by fetchKpi
export interface EvolutionKpi {
  emerging_count: number
  trust_mean: number
 /** D-cross: 与 /quality 共享 avg_skill_trust 对照口径（Neo4j Skill.trust_score 实时均值）*/
  trust_mean_neo4j_skill?: number
  cii_mean: number
  alert_count: number
  days: number
}

export const DEFAULT_KPI: EvolutionKpi = {
  emerging_count: 0,
  trust_mean: 0,
  trust_mean_neo4j_skill: 0,
  cii_mean: 0,
  alert_count: 0,
  days: 90,
}

// LOOP-06: Emerging alert type for evolution alerts
//: Added missing fields from backend (source_count, trend, portability_score)
export interface EmergingAlert {
  skill_name: string
  category: string
  level: 'emerging' | 'rising' | 'stable' | 'declining'
  z_score: number
  current_frequency: number
  mean_frequency: number
  source_count: number
  trend: 'rising' | 'stable' | 'declining'
  portability_score: number
  domains: string[]
  positions: string[]
  alert_message: string
}

// BUG-5 fix: review-queue item shape returned by /evolution/review-queue
export interface ReviewQueueItem {
 // E22 fix: include id so the frontend can dispatch per-row approve/reject
 // via /evolution/review-queue/{id}/action.
  id: string
  skill_name: string | null
  position_name: string | null
  change_type: string
  trust_score: number
  status: string
  created_at: string
}

// 快照联动: /evolution/cii-history/{position} 返回的 CII 历史（E1 快照时间线数据源）
export interface CiiHistoryEntry {
  snapshot_date: string
  cii: number
  total_skills: number
  inflated_skills: number
}

export const useEvolutionStore = defineStore('evolution', () => {
  const loading = ref(false)
  const trendItems = ref<TrendItem[]>([])

  const snapshotsLoading = ref(false)
  const snapshots = ref<SnapshotEntry[]>([])

  const changelogLoading = ref(false)
  const changelogData = ref<ChangelogEntry[]>([])

 // LOOP-06: Emerging alerts state
  const emergingAlerts = ref<EmergingAlert[]>([])
  const alertsLoading = ref(false)

 // 10-03 : KPI row state — zeros until fetchKpi resolves
  const kpi = ref<EvolutionKpi>({ ...DEFAULT_KPI })
  const kpiLoading = ref(false)

  async function fetchTrends(days?: number) {
    loading.value = true
    try {
      const params = days ? { days } : undefined
      const data = validateEvolution(
        await request.get<{ items?: TrendItem[] }>('/evolution/trends', { params }) as Record<string, unknown>,
        evolutionSchema, '/evolution/trends', 'EvolutionTrendsResponse',
      ) as { items?: TrendItem[] }
      trendItems.value = data.items ?? []
    } finally {
      loading.value = false
    }
    return { items: trendItems.value }
  }

 // 10-03 : KPI row fetch — mirrors fetchTrends pattern
  async function fetchKpi(days?: number) {
    kpiLoading.value = true
    try {
      const params = days ? { days } : undefined
      const data = validateEvolution(
        await request.get<{ emerging_count?: number; trust_mean?: number; trust_mean_neo4j_skill?: number; cii_mean?: number; alert_count?: number; days?: number }>(
          '/evolution/kpi', { params },
        ) as Record<string, unknown>,
        evolutionSchema, '/evolution/kpi', 'EvolutionKpiResponse',
      ) as { emerging_count?: number; trust_mean?: number; trust_mean_neo4j_skill?: number; cii_mean?: number; alert_count?: number; days?: number }
      kpi.value = {
        emerging_count: data.emerging_count ?? DEFAULT_KPI.emerging_count,
        trust_mean: data.trust_mean ?? DEFAULT_KPI.trust_mean,
        trust_mean_neo4j_skill: data.trust_mean_neo4j_skill ?? DEFAULT_KPI.trust_mean_neo4j_skill,
        cii_mean: data.cii_mean ?? DEFAULT_KPI.cii_mean,
        alert_count: data.alert_count ?? DEFAULT_KPI.alert_count,
        days: data.days ?? DEFAULT_KPI.days,
      }
    } finally {
      kpiLoading.value = false
    }
    return kpi.value
  }

 // 触发演化分析（原 EvolutionDashboard.vue 直调 request.post('/evolution/analyze')）
  async function analyze(days?: number): Promise<{ message?: string; task_id?: string; days?: number }> {
    const res = await request.post<{ message?: string; task_id?: string; days?: number }>(
      '/evolution/analyze', undefined, { params: days ? { days } : undefined },
    ) as { message?: string; task_id?: string; days?: number }
    return res
  }

  async function fetchSnapshots(limit = 50) {
    snapshotsLoading.value = true
    try {
 // P2 fix (functional-review 2026-08-13): 后端 /evolution/snapshots 返回
 // SnapshotEntry 数组，此前把整数组当单条 SnapshotEntry 校验 → schema 名
 // 错位，校验实际未生效（DEV warn 不报）。改为逐条校验。
      const data = await request.get(`/evolution/snapshots?limit=${limit}`) as SnapshotEntry[]
      const list = Array.isArray(data)
        ? data.map((item) => validateEvolution(item, evolutionSchema, '/evolution/snapshots', 'SnapshotEntry') as SnapshotEntry)
        : []
      snapshots.value = [...list].sort((a, b) =>
        String(a.snapshot_date).localeCompare(String(b.snapshot_date))
      ) as SnapshotEntry[]
    } catch {
      snapshots.value = []
    } finally {
      snapshotsLoading.value = false
    }
    return snapshots.value
  }

 // UX-04: Renamed parameter — backend 'identifier' accepts both position and skill names
  async function fetchChangelog(identifier: string) {
    changelogLoading.value = true
    try {
      const raw = validateEvolution(
        await request.get(`/evolution/changelog/${encodeURIComponent(identifier)}`) as Record<string, unknown>,
        evolutionSchema, `/evolution/changelog/${encodeURIComponent(identifier)}`, 'ChangelogEntry',
      ) as unknown
      if (raw === null || raw === undefined) {
        changelogData.value = []
        return changelogData.value
      }
      const data = raw as Record<string, unknown>
      changelogData.value = (Array.isArray(raw)
        ? raw
        : data.changelog ?? data.items ?? []) as ChangelogEntry[]
    } finally {
      changelogLoading.value = false
    }
    return changelogData.value
  }

 // LOOP-06: Fetch emerging skill alerts
  async function fetchEmergingAlerts(level?: string) {
    alertsLoading.value = true
    try {
      const params = level ? { level } : {}
      const data = validateEvolution(
        await request.get('/evolution/emerging-alerts', { params }) as { alerts: EmergingAlert[]; total: number; summary: string },
        evolutionSchema, '/evolution/emerging-alerts', 'EmergingAlertsResponse',
      ) as { alerts: EmergingAlert[]; total: number; summary: string }
      emergingAlerts.value = data.alerts ?? []
    } catch (e: unknown) {
      if (import.meta.env.DEV) console.error('[Evolution] Failed to fetch alerts:', e)
 // fix: HTTPException.detail 在 axios 错误对象里位于 response.data.detail
      const detail = (e as { response?: { data?: { detail?: string } } })?.response?.data?.detail
      if (detail) console.error('[Evolution] Detail:', detail)
      emergingAlerts.value = []
    } finally {
      alertsLoading.value = false
    }
  }

 // BUG-5 fix: low-trust EvolutionChangelog review queue
  const reviewQueue = ref<ReviewQueueItem[]>([])
  const reviewQueueLoading = ref(false)
  async function fetchReviewQueue(status: string = 'pending') {
    reviewQueueLoading.value = true
    try {
      const data = validateEvolution(
        await request.get(`/evolution/review-queue`, { params: { status } }) as ReviewQueueItem[],
        evolutionSchema, '/evolution/review-queue', 'ReviewQueueItem',
      ) as ReviewQueueItem[]
      reviewQueue.value = Array.isArray(data) ? data : []
    } catch (e: unknown) {
      if (import.meta.env.DEV) console.error('[Evolution] Failed to fetch review queue:', e)
      reviewQueue.value = []
    } finally {
      reviewQueueLoading.value = false
    }
    return reviewQueue.value
  }

 // 快照联动 (E1): 指定岗位的 CII 历史（真实快照数据，非估算）
  const ciiHistory = ref<CiiHistoryEntry[]>([])
  const ciiHistoryLoading = ref(false)
  const ciiHistoryPosition = ref('')
  async function fetchCiiHistory(position: string) {
    if (!position) { ciiHistory.value = []; ciiHistoryPosition.value = ''; return ciiHistory.value }
    ciiHistoryLoading.value = true
    ciiHistoryPosition.value = position
    try {
      const data = await request.get<{ position?: string; history?: CiiHistoryEntry[] }>(
        `/evolution/cii-history/${encodeURIComponent(position)}`,
      ) as { position?: string; history?: CiiHistoryEntry[] }
      ciiHistory.value = data.history ?? []
    } catch (e: unknown) {
      if (import.meta.env.DEV) console.error('[Evolution] Failed to fetch CII history:', e)
      ciiHistory.value = []
    } finally {
      ciiHistoryLoading.value = false
    }
    return ciiHistory.value
  }

 // 10-03 : manual refresh — fire all dashboard fetches concurrently
  async function refreshAll() {
    await Promise.all([
      fetchTrends(),
      fetchSnapshots(),
      fetchEmergingAlerts(),
      fetchKpi(),
    ])
  }

  return {
    loading,
    trendItems,
    snapshotsLoading,
    snapshots,
    changelogLoading,
    changelogData,
    emergingAlerts,
    alertsLoading,
    kpi,
    kpiLoading,
    reviewQueue,
    reviewQueueLoading,
    ciiHistory,
    ciiHistoryLoading,
    ciiHistoryPosition,
    fetchTrends,
    fetchKpi,
    analyze,
    fetchSnapshots,
    fetchChangelog,
    fetchEmergingAlerts,
    fetchReviewQueue,
    fetchCiiHistory,
    refreshAll,
  }
})

File: backend/app/schemas/evolution.py
"""演化域 Schema：职业路径/预警/行业报告/演化趋势 (PLAN-014 批次10-12 迁入集中管理)。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class CareerPathNode(BaseModel):
    """A node in the career path graph."""

    position: str = Field(..., description="岗位名称")
    similarity: float = Field(default=0.0, ge=0, le=1, description="与起点岗位的相似度 0~1")
    skill_overlap: list[str] = Field(default_factory=list, description="技能重叠列表")
    key_gaps: list[str] = Field(default_factory=list, description="关键技能缺口")
    evidence_count: int = Field(default=0, ge=0, description="路径证据条数")
    direction: str = Field(default="forward", description="forward | lateral | up")


class CareerPathResponse(BaseModel):
    """Career path planning response."""

    origin: str
    nodes: list[CareerPathNode] = Field(default_factory=list)
    total_paths: int = 0


class EmergingAlert(BaseModel):
    """An emerging skill alert with full context."""

    skill_name: str = Field(..., description="技能名称")
    category: str = Field(default="", description="分类")
    level: str = Field(..., description="分类: emerging/rising/declining/stable")
    z_score: float = Field(default=0.0, description="Z-score 值")
    current_frequency: int = Field(default=0, description="当前频次")
    mean_frequency: float = Field(default=0.0, description="历史均值频次")
    source_count: int = Field(default=0, description="来源数")
    domains: list[str] = Field(default_factory=list, description="所属领域")
    positions: list[str] = Field(default_factory=list, description="关联岗位")
    trend: str = Field(default="stable", description="趋势方向")
    portability_score: float = Field(default=0.0, ge=0, le=1, description="可迁移性得分")
    alert_message: str = Field(default="", description="预警描述")


class EmergingAlertsResponse(BaseModel):
    """Emerging skill alerts response."""

    alerts: list[EmergingAlert] = Field(default_factory=list, description="预警列表")
    total: int = 0
    summary: str = ""


class SkillTrendItem(BaseModel):
    """Skill trend in industry report."""

    skill_name: str = Field(..., description="技能名称")
    trend: str = Field(..., description="rising | stable | declining")
    frequency: int = Field(default=0, ge=0, description="出现频次")
    source_count: int = Field(default=0, ge=0, description="数据源数")
    related_positions: list[str] = Field(default_factory=list, description="相关岗位")


class IndustryReportResponse(BaseModel):
    """Industry trend report response."""

    total_skills: int = Field(default=0, ge=0, description="技能总数")
    rising_skills: list[SkillTrendItem] = Field(default_factory=list, description="上升技能")
    declining_skills: list[SkillTrendItem] = Field(default_factory=list, description="下降技能")
    stable_skills: list[SkillTrendItem] = Field(default_factory=list, description="平稳技能")
    top_positions: list[dict] = Field(default_factory=list, description="热门岗位 TOP")
    summary: str = Field(default="", description="报告摘要")


class EvolutionTrend(BaseModel):
    """技能趋势条目。"""

    skill_name: str = Field(..., description="技能名称")
    trend: str = Field(..., description="趋势方向：rising/stable/declining")
    confidence: float = Field(..., ge=0, le=1, description="趋势置信度")
    points: list[float] = Field(default_factory=list, description="CII 时序数据点")
    related_positions: list[str] = Field(default_factory=list, description="相关岗位")


class EvolutionTrendsResponse(BaseModel):
    """演化趋势响应。"""

    items: list[EvolutionTrend] = Field(default_factory=list, description="趋势列表")


class ChangelogEntry(BaseModel):
    """变更日志条目。"""

    id: str
    skill_name: str
    change_type: str
    old_proficiency: str | None = None
    new_proficiency: str | None = None
    old_requirement: str | None = None
    new_requirement: str | None = None
    trust_score: float
    confidence: float
    created_at: datetime
    position_name: str = Field(default="", description="发生技能变更的职位名称")
    status: str = Field(default="pending", description="审核状态：pending/approved/rejected")
    written_back: bool = Field(default=False, description="该变更是否已回写 position_skill_relations")
    evidence_json: dict[str, Any] = Field(
        default_factory=dict,
        description="证据链：mention_count_old/new、source_count、factors（源计数/提及新旧/变更类型/稳定性因子）",
    )


class EvolutionKpiResponse(BaseModel):
    """演化看板 KPI 行响应。"""

    emerging_count: int = Field(..., ge=0, description="涌现技能数（emerging + rising）")
    # 2026-08-21 (debug 修复): Optional —— changelog 空表时返回 None（前端显示"—"），
    # 而非 0.0（用户误读为"信任度全 0"）
    trust_mean: float | None = Field(default=None, ge=0, le=1, description="变更日志信任度均值（0-1，真实聚合；空表为 None）")
    trust_mean_neo4j_skill: float = Field(
        default=0.0, ge=0, le=1, description="Neo4j Skill.trust_score 实时均值（与 /quality 共享 avg_skill_trust 指标）"
    )
    cii_mean: float = Field(..., ge=0, description="技能 CII 均值（基准 100，A4 基线口径）")
    alert_count: int = Field(..., ge=0, description="预警数（emerging/rising/declining 非平稳信号）")
    days: int = Field(..., ge=7, le=730, description="分析时间窗口（天）")


class EvolutionPathEntry(BaseModel):
    """演化路径条目。"""

    id: str
    source_position: str
    target_position: str
    similarity: float
    evidence_count: int
    skill_overlap: list[str]
    key_gaps: list[str]
    trust_score: float
    trend: str = "stable"


class EmergingSkill(BaseModel):
    """涌现技能条目。"""

    skill_name: str
    level: str  # emerging/rising/stable/declining
    z_score: float
    current_frequency: int
    mean_frequency: float
    source_count: int
    positions: list[str]


class SnapshotEntry(BaseModel):
    """快照条目。"""

    id: str
    position_name: str
    snapshot_date: datetime
    required_skills: list[dict[str, Any]]
    preferred_skills: list[dict[str, Any]]
    source_count: int


class ReviewQueueItem(BaseModel):
    """审核队列条目。"""

 # E22 fix: include id so the frontend can dispatch per-row approve/reject
 # via /evolution/review-queue/{id}/action.
    id: str = Field(..., description="EvolutionChangelog id (UUID)")
    skill_name: str
    position_name: str
    change_type: str
    trust_score: float
    status: str  # pending/approved/rejected
    created_at: datetime


# ─── Endpoints ───


class PortabilityDetail(BaseModel):
    """Skill portability analysis response."""

    skill_name: str = Field(..., description="技能名称")
    portability_score: float = Field(default=0.0, ge=0, le=1, description="可迁移性得分")
    domains: list[str] = Field(default_factory=list, description="所属领域")
    domain_count: int = 0
    positions_by_domain: dict[str, list[str]] = Field(
        default_factory=dict,
        description="各领域关联岗位",
    )
    total_positions: int = 0
    transferability_tier: str = Field(default="low", description="可迁移性等级")
    related_skills: list[str] = Field(default_factory=list, description="相关跨领域技能")
    recommendation: str = Field(default="", description="建议")

class CausalAssociation(BaseModel):
    """技能-岗位关联显著性 (§7.6 因果推理轻量版)。"""

    position: str = Field(..., min_length=1, max_length=200, description="岗位名称")
    a: int = Field(..., ge=0, description="技能记录中含该岗位的记录数")
    b: int = Field(..., ge=0, description="技能记录中不含该岗位的记录数")
    c: int = Field(..., ge=0, description="对照记录中含该岗位的记录数")
    d: int = Field(..., ge=0, description="对照记录中不含该岗位的记录数")
    p_value: float = Field(..., ge=0, le=1, description="Fisher 精确检验 p 值")
    significant: bool = Field(..., description="p < 0.05 且 |phi| >= 0.1")
    phi: float = Field(..., ge=-1, le=1, description="phi 系数 (效应量)")
    method: str = Field(default="fisher_exact", description="检验方法")


class CausalAnalysisResponse(BaseModel):
    """技能因果关联分析响应 (§7.6)。"""

    skill: str = Field(..., min_length=1, description="被分析技能")
    associations: list[CausalAssociation] = Field(default_factory=list, description="显著关联岗位列表")
    total_records: int = Field(0, ge=0, description="技能记录数")
    control_records: int = Field(0, ge=0, description="对照记录数")
    alpha: float = Field(0.05, ge=0, le=1, description="显著性水平")
